internal/wsutil/wsutil.go

```diff
@@ -109,48 +81,61 @@ func WebsocketSendStream(conn MessageWriter, r io.Reader, bufferSize int) chan b
 func WebsocketRecvStream(w io.Writer, conn MessageReader) chan bool {
 	ch := make(chan bool)
 
-	go func(w io.Writer, conn MessageReader) {
-		for {
-			mt, r, err := conn.NextReader()
-			if mt == websocket.CloseMessage {
-				logger.Debugf("Got close message for reader")
-				break
-			}
-
-			if mt == websocket.TextMessage {
-				logger.Debugf("Got message barrier")
-				break
-			}
+	go func() {
+		recvLoop(w, conn)
+		close(ch)
+	}()
 
-			if err != nil {
-				logger.Debugf("Got error getting next reader %s", err)
-				break
-			}
+	return ch
+}
 
-			buf, err := ioutil.ReadAll(r)
-			if err != nil {
-				logger.Debugf("Got error writing to writer %s", err)
-				break
-			}
+func recvLoop(w io.Writer, conn MessageReader) {
+	for {
+		mt, r, err := conn.NextReader()
+		if err != nil {
+			logger.Debugf("Cannot get next reader: %v", err)
+			return
+		}
 
-			if w == nil {
-				continue
-			}
+		if mt == websocket.CloseMessage {
+			logger.Debugf("Got close message for reader")
+			return
+		}
 
-			i, err := w.Write(buf)
-			if i != len(buf) {
-				logger.Debugf("Didn't write all of buf")
-				break
+		if mt == websocket.TextMessage {
+			var command struct {
+				Command string `json:"command"`
 			}
+			err := json.NewDecoder(r).Decode(&command)
```

Comment:
What's the intended handling of the remaining data in the reader here?  I suppose right now it's ignoring anything after the first JSON object.  We probably don't care much for now while we really only have one message and at the very end, but I'm curious if there's candidate bug here or not. If so, there should at least be a note here to remind ourselves that this is an issue.

Reply:
Yeah, fair point. I've changed to use `json.Unmarshal` which only supports a single value. `json.Decoder` is more for streams in any case.